# Nombre: Andrés Carrillo

## Desarrollo

1. Preprocesamiento

In [4]:
!pip install nltk

Defaulting to user installation because normal site-packages is not writeable
  Using cached nltk-3.9.1-py3-none-any.whl.metadata (2.9 kB)
  Using cached click-8.2.1-py3-none-any.whl.metadata (2.5 kB)
Using cached nltk-3.9.1-py3-none-any.whl (1.5 MB)
Using cached click-8.2.1-py3-none-any.whl (102 kB)



[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
import pandas as pd
import json
import re
import string
import nltk
nltk.download('punkt')
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from tqdm import tqdm

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\andre\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\andre\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
with open('arxiv-metadata-oai-snapshot.json', 'r') as f:
    lines = [json.loads(line) for _, line in zip(range(int(0.01 * 1700000)), f)]

df = pd.DataFrame(lines)
df = df[['id', 'title', 'abstract', 'authors', 'categories', 'update_date']]

# Preprocesamiento
stop_words = set(stopwords.words('english'))
def preprocess(text):
    text = text.lower()
    text = re.sub(f'[{re.escape(string.punctuation)}]', '', text)
    tokens = word_tokenize(text)
    tokens = [t for t in tokens if t not in stop_words]
    return ' '.join(tokens)

tqdm.pandas()
df['text'] = (df['title'] + ' ' + df['abstract']).progress_apply(preprocess)

100%|██████████| 17000/17000 [00:06<00:00, 2501.49it/s]


In [14]:
df

id                                              title  \
0      0704.0001  Calculation of prompt diphoton production cros...   
1      0704.0002           Sparsity-certifying Graph Decompositions   
2      0704.0003  The evolution of the Earth-Moon system based o...   
3      0704.0004  A determinant of Stirling cycle numbers counts...   
4      0704.0005  From dyadic $\Lambda_{\alpha}$ to $\Lambda_{\a...   
...          ...                                                ...   
16995  0707.3825  Emergence of noncollinear magnetic ordering in...   
16996  0707.3826                      More hilltop inflation models   
16997  0707.3827  Engineering Silicon Nanocrystals: Theoretical ...   
16998  0707.3828  Structure, bonding and magnetism in cobalt clu...   
16999  0707.3829  Occupation Statistics of Critical Branching Ra...   

                                                abstract  \
0        A fully differential calculation in perturba...   
1        We describe a new algorithm, the $(k,\ell)$-...   
2        The evolution of Earth-Moon system is descri...   
3        We show that a determinant of Stirling cycle...   
4        In this paper we show how to compute the $\L...   
...                                                  ...   
16995    Using first-principles density functional ca...   
16996    Using analytic expressions, we explore the p...   
16997    We show that the optical and electronic prop...   
16998    The structural, electronic and magnetic prop...   
16999    Consider a critical nearest neighbor branchi...   

                                                 authors  \
0      C. Bal\'azs, E. L. Berger, P. M. Nadolsky, C.-...   
1                        Ileana Streinu and Louis Theran   
2                                            Hongjun Pan   
3                                           David Callan   
4               Wael Abu-Shammala and Alberto Torchinsky   
...                                                  ...   
16995   Mukul Kabir, D. G. Kanhere and Abhijit Mookerjee   
16996        Kazunori Kohri, Chia-Min Lin, David H. Lyth   
16997  Federico Iori, Elena Degoli, Rita Magri, Ivan ...   
16998  S. Datta, M. Kabir, S. Ganguly, B. Sanyal, T. ...   
16999                       Steven Lalley, Xinghua Zheng   

                             categories update_date  \
0                                hep-ph  2008-11-26   
1                         math.CO cs.CG  2008-12-13   
2                        physics.gen-ph  2008-01-13   
3                               math.CO  2007-05-23   
4                       math.CA math.FA  2013-10-15   
...                                 ...         ...   
16995                  physics.atm-clus  2009-11-13   
16996            hep-ph astro-ph hep-th  2008-11-26   
16997  cond-mat.other cond-mat.mtrl-sci  2009-11-13   
16998                  physics.atm-clus  2015-05-13   
16999                           math.PR  2010-04-08   

                                                    text  
0      calculation prompt diphoton production cross s...  
1      sparsitycertifying graph decompositions descri...  
2      evolution earthmoon system based dark matter f...  
3      determinant stirling cycle numbers counts unla...  
4      dyadic lambdaalpha lambdaalpha paper show comp...  
...                                                  ...  
16995  emergence noncollinear magnetic ordering small...  
16996  hilltop inflation models using analytic expres...  
16997  engineering silicon nanocrystals theoretical s...  
16998  structure bonding magnetism cobalt clusters st...  
16999  occupation statistics critical branching rando...  

[17000 rows x 7 columns]

2. Indexación

In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(df['text'])

In [15]:
tfidf_matrix

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 983235 stored elements and shape (17000, 71211)>

BM25

In [11]:
!pip install rank_bm25

Defaulting to user installation because normal site-packages is not writeable
  Using cached rank_bm25-0.2.2-py3-none-any.whl.metadata (3.2 kB)
Using cached rank_bm25-0.2.2-py3-none-any.whl (8.6 kB)



[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [12]:
from rank_bm25 import BM25Okapi

tokenized_corpus = [doc.split() for doc in df['text']]
bm25 = BM25Okapi(tokenized_corpus)

Embeddings + FAISS

In [16]:
from sentence_transformers import SentenceTransformer
import faiss
import numpy as np

In [17]:
model = SentenceTransformer('all-MiniLM-L6-v2')

In [19]:
pip install sentence-transformers faiss-cpu numpy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [20]:
embeddings = model.encode(df['text'].tolist(), show_progress_bar=True)
embeddings = np.array(embeddings).astype('float32')

index = faiss.IndexFlatL2(embeddings.shape[1])
index.add(embeddings)

Batches: 100%|██████████| 532/532 [03:51<00:00,  2.29it/s]


3. Recuperación

In [21]:
def search_tfidf(query, top_k=10):
    q_vec = tfidf_vectorizer.transform([preprocess(query)])
    scores = (tfidf_matrix @ q_vec.T).toarray().ravel()
    top_idx = scores.argsort()[-top_k:][::-1]
    return df.iloc[top_idx][['id', 'title', 'abstract']]

def search_bm25(query, top_k=10):
    q_tok = preprocess(query).split()
    scores = bm25.get_scores(q_tok)
    top_idx = np.argsort(scores)[-top_k:][::-1]
    return df.iloc[top_idx][['id', 'title', 'abstract']]

def search_faiss(query, top_k=10):
    q_emb = model.encode([preprocess(query)]).astype('float32')
    D, I = index.search(q_emb, top_k)
    return df.iloc[I[0]][['id', 'title', 'abstract']]

4. RAG 

In [6]:
import os
from langchain.chains import RetrievalQA
from langchain_google_genai import ChatGoogleGenerativeAI

In [ ]:

os.environ["GOOGLE_API_KEY"] = "****"

llm = ChatGoogleGenerativeAI(model="models/gemini-1.5-flash", temperature=0)
def rag_answer(query):
    docs = search_faiss(query, top_k=3)
    context = "\n\n".join(docs['abstract'].tolist())
    prompt = f"""Contexto:\n{context}\n\nPregunta:\n{query}\n\nResponde en español, resume la información y justifica la relevancia de los documentos."""
    return llm.invoke(prompt)

5. Evaluación

In [24]:
import pandas as pd

In [25]:
with open("queries.txt", "r", encoding="utf-8") as f:
    queries = [line.strip() for line in f if line.strip()]

resultados = []

In [28]:
for query in queries:
    print(f"\n=== Consulta: {query} ===")
    tfidf_res = search_tfidf(query)
    bm25_res = search_bm25(query)
    faiss_res = search_faiss(query)
    rag_res = rag_answer(query)

    print("TF-IDF:")
    print(tfidf_res[['id', 'title']])
    print("BM25:")
    print(bm25_res[['id', 'title']])
    print("FAISS:")
    print(faiss_res[['id', 'title']])
    print("Respuesta del RAG:")
    print(rag_res)

    # Medir similitud entre rankings
    tfidf_ids = set(tfidf_res['id'])
    bm25_ids = set(bm25_res['id'])
    faiss_ids = set(faiss_res['id'])
    interseccion = tfidf_ids & bm25_ids & faiss_ids

    print("Coincidencias top-10 entre los tres modelos:", len(interseccion))

    resultados.append({
        "consulta": query,
        "tfidf_top1": tfidf_res.iloc[0]['title'],
        "bm25_top1": bm25_res.iloc[0]['title'],
        "faiss_top1": faiss_res.iloc[0]['title'],
        "coincidencias_top10": len(interseccion),
        "rag_respuesta": rag_res
    })



=== Consulta: diphoton production cross sections ===
TF-IDF:
              id                                              title
7806   0705.3804  Exclusive e+e-, Di-photon and Di-jet Productio...
0      0704.0001  Calculation of prompt diphoton production cros...
15464  0707.2294  Search for a High-Mass Diphoton State and Limi...
8315   0705.4313  Projectile Fragmentation of $^{86}$Kr at 64 Me...
11499  0706.2813  Measurement of the Total Hadronic Cross Sectio...
10803  0706.2117  Novel Master Formula for Twist-3 Soft-Gluon-Po...
4351   0705.0349  New isotope 44Si and systematics of the produc...
11379  0706.2693  Reaction cross sections for proton scattering ...
5053   0705.1051  Quasi-elastic neutrino charged-current scatter...
12703  0706.4017  The Role of Gauge Invariance in Single-Spin As...
BM25:
              id                                              title
7806   0705.3804  Exclusive e+e-, Di-photon and Di-jet Productio...
0      0704.0001  Calculation of prompt diphoton

In [27]:
resultados

[{'consulta': 'diphoton production cross sections',
  'tfidf_top1': 'Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron',
  'bm25_top1': 'Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron',
  'faiss_top1': 'Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron',
  'coincidencias_top10': 3,
  'rag_respuesta': AIMessage(content='Los documentos presentan información sobre las secciones transversales de producción de difotones en colisiones hadrónicas, con énfasis en diferentes aspectos y energías.\n\n**Resumen:**\n\n* **Documento 1:** Se centra en la observación y medición de la sección transversal de la producción exclusiva de pares e+e- y di-jets en colisiones protón-antiprotón en el Tevatron de Fermilab.  Aunque no se centra directamente en difotones,  establece un contexto experimental importante para la producción de pares de partículas en colisiones hadrónicas.\n\n* **Documento 2:** Presenta un cálculo completamente diferencial en QCD perturbativ

## 6. Informe

### Implementación de la arquitectura
La arquitectura implementada incluye:
- Preprocesamiento del corpus (minúsculas, stopwords, tokenización)
- Indexación con TF-IDF, BM25 y FAISS (embeddings)
- Funciones de búsqueda para cada modelo
- Módulo RAG integrado con Gemini

### Tabla comparativa de resultados entre modelos


| Consulta                               | TF-IDF Top 1                                                                 | BM25 Top 1                                                                 | FAISS Top 1                                                                | Coincidencias Top-10 |
|----------------------------------------|------------------------------------------------------------------------------|----------------------------------------------------------------------------|---------------------------------------------------------------------------|----------------------|
| diphoton production cross sections     | Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron              | Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron             | Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron            | 3                    |
| quantum chromodynamics                 | Monte Carlo Methods in Quantum Field Theory                                  | Two interacting GL-equations in High-T$_c$ superconductivity and quantum chromodynamics | Monte Carlo Methods in Quantum Field Theory                                | 3                    |
| higgs boson decay                      | Prospects for the search for Higgs bosons with vector boson fusion processes at the LHC | Invisibly decaying Higgs boson in the Littlest Higgs model with T-parity    | Invisibly decaying Higgs boson in the Littlest Higgs model with T-parity   | 3                    |
| machine learning for particle physics  | An Adaptive Strategy for the Classification of G-Protein Coupled Receptors   | An Adaptive Strategy for the Classification of G-Protein Coupled Receptors  | Bayesian Learning of Neural Networks for Signal/Background Discrimination in Particle Physics | 1                    |
| top quark production                   | Measurements of Top Properties at the Tevatron                               | Measurements of Top Properties at the Tevatron                              | Top Quark Production and Decay Properties at the Tevatron                  | 2                    |

### Ejemplo de una consulta y su respuesta generada con RAG

**Consulta:** machine learning for particle physics

**Respuesta RAG:**

In [30]:
respuesta_rag = rag_answer("machine learning for particle physics")
print(respuesta_rag)

content='Los documentos presentados ilustran el uso del aprendizaje automático, específicamente de redes neuronales, en la física de partículas.  Se destacan dos enfoques:\n\n* **Redes Neuronales Bayesianas para Discriminación de Señales:** Un documento describe el uso de redes neuronales bayesianas para la discriminación entre señal y ruido en la búsqueda de leptoquarks de segunda generación en el Tevatron.  Se compara el rendimiento de este método bayesiano con el entrenamiento convencional de redes neuronales, mostrando la superioridad del enfoque bayesiano en términos de optimización y robustez.  Esto es relevante porque la clasificación precisa de eventos es crucial en la física de partículas experimental para identificar nuevas partículas o fenómenos.\n\n* **Métodos computacionales eficientes para simulaciones:** Otro documento presenta un método eficiente para calcular la energía electrostática y las fuerzas entre partículas, considerando condiciones de contorno específicas (per

### Diferencias entre modelos y utilidad del RAG

1. ¿Cuáles documentos aparecen en común?

Los tres modelos coinciden en algunos documentos relevantes. Por ejemplo, para la consulta "diphoton production cross sections" el documento 0705.3804 aparece en el top-1 de los tres modelos, esto podría indicarme que, aunque los metodos de recuperación son diferentes existen documentos que son realmente relevantes como en este caso.

In [48]:
for r in resultados:
    if r["consulta"] == "diphoton production cross sections":
        print("TF-IDF Top 1:", r["tfidf_top1"])
        print("BM25 Top 1:", r["bm25_top1"])
        print("FAISS Top 1:", r["faiss_top1"])

TF-IDF Top 1: Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron
BM25 Top 1: Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron
FAISS Top 1: Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron
TF-IDF Top 1: Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron
BM25 Top 1: Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron
FAISS Top 1: Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron


Además según la tabla comparativa el promedio de las coincidencias top-10 es de 3 documentos presentes en los 3 modelos al mismo tiempo

In [44]:
query = "diphoton production cross sections"
tfidf_res = search_tfidf(query)
bm25_res = search_bm25(query)
faiss_res = search_faiss(query)

tfidf_ids = set(tfidf_res['id'])
bm25_ids = set(bm25_res['id'])
faiss_ids = set(faiss_res['id'])
comunes = tfidf_ids & bm25_ids & faiss_ids

print("Documentos en común en el top-10 de los tres modelos:")
for doc_id in comunes:
    titulo = df[df['id'] == doc_id]['title'].values[0]
    print(f"- {doc_id}: {titulo}")
print(f"Total de coincidencias top-10: {len(comunes)}")

Documentos en común en el top-10 de los tres modelos:
- 0705.4313: Projectile Fragmentation of $^{86}$Kr at 64 MeV/nucleon
- 0704.0001: Calculation of prompt diphoton production cross sections at Tevatron and
  LHC energies
- 0705.3804: Exclusive e+e-, Di-photon and Di-jet Production at the Tevatron
Total de coincidencias top-10: 3


2. ¿Qué diferencias hay en el ordenamiento?

Aunque hay coincidencias en los documentos recuperados, el orden en que aparecen varía entre modelos. Por ejemplo, para la consulta "top quark production", el documento 0706.1984 es el primero tanto en TF–IDF como en BM25, pero en FAISS aparece en segunda posición. 

Además, FAISS tiende a incluir algunos artículos que no aparecen en los otros modelos, ya que prioriza la similitud semántica, mientras que TF–IDF y BM25 dependen más de la frecuencia y el peso de los términos exactos.

In [52]:
for r1 in resultados:
    if r1["consulta"] == "top quark production":
        print("TF-IDF Top 1:", r1["tfidf_top1"])
        print("BM25 Top 1:", r1["bm25_top1"])
        print("FAISS Top 1:", r1["faiss_top1"])

TF-IDF Top 1: Measurements of Top Properties at the Tevatron
BM25 Top 1: Measurements of Top Properties at the Tevatron
FAISS Top 1: Top Quark Production and Decay Properties at the Tevatron
TF-IDF Top 1: Measurements of Top Properties at the Tevatron
BM25 Top 1: Measurements of Top Properties at the Tevatron
FAISS Top 1: Top Quark Production and Decay Properties at the Tevatron
